### ISSUE TITLE
Not easy enough to redispatch steps

### ISSUE THREAD

**Issue by pjlsergeant** [pjlsergeant/test-bdd-cucumber-perl#28]
It should be easy to redispatch steps from inside a step definition:

```perl
Given "something", sub {
  C->run( Given => "I did setup step 1");
  C->run(Given => "I did setup step 2");
  S->{'foo'} = S->{'bar'}->baz;
}
```

**Comment by kidddw** [comment]
So is there a way to call a step from a step?    I'd like to write a step that calls another step, but if not I can re-write my step, but didn't know if this was something that could be done.

**Comment by sheriff** [comment]
It's not something that can be done easily. I feel like it's something that could be implemented without too much trouble, but there are a lot of non-obvious design choices to make. I could rip out the `for` loop around:

https://github.com/sheriff/test-bdd-cucumber-perl/blob/master/lib/Test/BDD/Cucumber/Executor.pm#L230

and use that. Buuuuuuut: what do we do about the step context? Do we create a whole new one? Do test failures happen in this step or a sub-step? So I think it might be worth rewriting your step for now :-)

**Comment by neithere** [comment]
This is a really important thing, IMO.

We write *functional* tests using Robot Framework which is based on the idea of building "keywords" (steps) on top of other keywords. The lowest level is implemented in Python but there are usually many domain-specific, then project-specific, then module-specific levels upon it.

In this library for *unit* testing it would be enough to reuse steps from steps in *any* way; it's not the central concept here and I think the exact behaviour is not that important.

(By the way, thanks for this excellent library, it's far better than all other Perl alternatives; looks like it's the only one that doesn't miss the whole point of BDD and Cucumber.)

**Comment by chylli** [comment]
I also need this feature.

**Comment by pjlsergeant** [comment]
My reticence here is purely not liking any of the ways I can think of for how this would work; I would _love_ to hear you jump in with examples of how you'd use such a thing, and ideas for what it should look like in practice. Off the top of my head:

* If I redispatch a step, do I display your redispatch as a new line of output? Or is that hidden?

* Do failures in the redispatched step make the dispatching step fail?

* How much of the StepContext should be passed through? A whole new StepContext, or just a copy of the current one?

Not having a personal use-case (yet) for this, it's hard for me to know how you want this to look.

**Comment by chylli** [comment]
1. I hope it should be hidden when the sub-step success.
2. I think the dispatching step should fail. And the output of the sub-step should be included.
3. I'm not sure about this problem. But I think we can consider the called step as a called subroutine (procedure or function). So, maybe new StepContext?

At 2015-08-24 10:48:34, "Peter Sergeant" <notifications@github.com> wrote:


My reticence here is purely not liking any of the ways I can think of for how this would work; I would love to hear you jump in with examples of how you'd use such a thing, and ideas for what it should look like in practice. Off the top of my head:

If I redispatch a step, do I display your redispatch as a new line of output? Or is that hidden?

Do failures in the redispatched step make the dispatching step fail?

How much of the StepContext should be passed through? A whole new StepContext, or just a copy of the current one?

Not having a personal use-case (yet) for this, it's hard for me to know how you want this to look.

**Comment by neithere** [comment]
I mostly agree with @chylli.

> If I redispatch a step, do I display your redispatch as a new line of output? Or is that hidden?

It would be nice if it was hidden by default but visible in some verbose mode. In general, if everything goes fine, we don't need to see the inner logic, just the test itself. However, if the inner step fails, a detailed stack trace would be very helpful.

> Do failures in the redispatched step make the dispatching step fail?

It should definitely fail. Something like Robot Framework's "run keyword and ignore failure" / "...expect failure" would be nice though (i.e. the error is intercepted and converted to a return value). But it could be just an addition.

> How much of the StepContext should be passed through? A whole new StepContext, or just a copy of the current one?

Not sure. I guess we need real-life examples indeed. Here's one:

```cucumber
Scenario: ...
    Given search criterion X
    And search criterion Y
    And candidate A with attribute Y
    When search is done
    Then candidate A should have score 50%
```
This is fine until we need a matrix of combinations with a different number of criteria and attributes per row. The easiest thing is to provide a higher-level step `search criteria _, _, ...` instead of `search criterion _` and so on:
```cucumber
Scenario:
    Given search criteria <criteria>
    And candidate A with attributes <attributes>
    When search is done
    Then candidate A should have score <score>
    Examples:
        | criteria | attributes | score
        | X        | X          | 100%
        | Y        | X          | N/A
        | X, Y     | X, Z       | 50%
```
I'd love to just write something like this:
```perl
Given qr/^search criteria (.+?)$/, sub {
    S->given("search criterion $_") for split ', ', $1;
} 
```
...instead of repeating the code or extracting it from all steps to additional helpers.

**Comment by chylli** [comment]
Any progress ?

**Comment by pjlsergeant** [comment]
None, but looking at it again, I think I can see how I want this to work, which is 90% of the battle.

Firstly, add some kind of `clone` method to `Test::BDD::Cucumber::StepContext` which does an appropriate clone of the step context, and add a weak reference to the executor in the step context so it can handle a `->redispatch` call.

Secondly, add a `substep` argument to `find_and_dispatch` and `dispatch`, which changes their behaviour to `die` on any sort of error. This will get caught by the parent's `eval` in `dispatch` and lead to a failing step.

Finally, add `substep_x` methods to the base harness, so you can capture those if you want.

If you want to pass messages from your substep to your dispatching step, you need to push those in to the scenario-level stash.

How does that sound as a proposal?

**Comment by chylli** [comment]
I didn't read that source code. I will give you some advice after read it.

BTW, can the subroutine accept @_ also ? so that I can write like this before we have this feature:
```
Given qr/(a|b) step1/, \&step1;
Given qr/step2/,sub{
  step1('a');
}

sub step1{
    my $arg1 = shift;
}

```

**Comment by pjlsergeant** [comment]
I'd see it working more like:

@@@
Given qr/(a|b) step1/, \&step1;
Given qr/step2/,sub{
  C->redispatch("a step1");
}

sub step1 {
    warn "Argument was $1";
}
@@@

**Comment by pjlsergeant** [comment]
OK, there's an implementation here:

https://github.com/pjlsergeant/test-bdd-cucumber-perl/tree/step-redispatch

Docs are here:

https://github.com/pjlsergeant/test-bdd-cucumber-perl/blob/step-redispatch/lib/Test/BDD/Cucumber/StepContext.pm#L276

Please download it, and try it out!

**Comment by chylli** [comment]
I didn't use it yet. But can I give you 2 advice ?
1. dispatch is a little to long, maybe step is better.
2. Needn't separate 'verb' with 'text', writing the step text directly is more natural:
`C->step("Given I have pressed $1");`
the verb can be parsed by 'step' function

**Comment by pjlsergeant** [comment]
1. Yeah fine
2. The reason it doesn't do this is that there's no access to the parser
object at this point. Perhaps I don't need the whole thing, and can just
try and work out the language and use the verb keywords... Hrm

**Comment by neithere** [comment]
Cool! Looks really promising. Will try it a bit later. I would also prefer either of these:
```perl
    C->step("Given I have pressed $1");
    C->given("I have pressed $1");
```
instead of:
```perl
    C->dispatch("Given", "I have pressed $1");
```
However, it's not that important.

**Comment by pjlsergeant** [comment]
Has anyone had a play with it in its current form yet?

**Comment by pjlsergeant** [comment]
I gave this some more thought, and rereleased in its current form.

I specifically wanted to avoid having method names like the function names for creating steps - I've already had one confused email about it. So it's a bit more typing, but there you go. Anyway, merged and in 0.40.
